File: code-examples/examples/example-3/src/main/java/be/swsb/typesafety/example3/domain/booking/Booking.java
package be.swsb.typesafety.example3.domain.booking;

import java.util.UUID;

public class Booking {
    private UUID id;
    private BookingCode bookingCode;

    //imagine a bunch of other fields

    private Booking(final BookingCode bookingCode) {
        this.id = UUID.randomUUID();
        this.bookingCode = bookingCode;
    }

    public static Booking booking(final BookingCode bookingCode) {
        return new Booking(bookingCode);
    }

    public BookingCode getBookingCode() {
        return bookingCode;
    }

    public String getBookingCodeAsString() {
        return bookingCode.asString();
    }

    public UUID getId() {
        return id;
    }
}

File: code-examples/examples/example-5/src/main/java/be/swsb/typesafety/example5/domain/booking/Booking.kt
package be.swsb.typesafety.example5.domain.booking

import java.time.LocalDateTime
import java.util.*

class Booking(val id: UUID = UUID.randomUUID(),
              val bookingCode: BookingCode? = null,
              val reservationDate: LocalDateTime = LocalDateTime.now()) {
    //imagine a bunch of other fields

    fun getBookingCodeAsString(): String? {
        return bookingCode?.asString()
    }

    companion object {
        fun booking(bookingCode: BookingCode?): Booking {
            return Booking(bookingCode = bookingCode)
        }
    }
}
